### PR TITLE
Safe cluster state task notifications

### PR DESCRIPTION
**Pull request by jasontedor**
This commit addresses an issue where a cluster state task listener
throwning an exception could prevent other listeners from being
notified, and could prevent the executor from receiving notifications
that a new cluster state was published. Additionally, this commit also
addresses a similar issue for executors handling cluster state
publication notifications.

### CODE REVIEW

**Comment by jasontedor** [comment]
Adding back the v2.2.0 label.

File: core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java [hunk 1]
```diff
@@ -628,7 +629,61 @@ public String source() {
...
+            });
+    }
+
+    private void safely(Runnable runnable, Consumer<Exception> consumer) {
+        try {
+            runnable.run();
+        } catch (Exception e) {
+            try {
+                consumer.accept(e);
+            } catch (Exception ignored) {}
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: It feels squicky to ignore the exception. Personally I'd `assert false : "Exceptions not expected here."` and `logger.error` about it.

File: core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java [hunk 2]
```diff
@@ -628,7 +629,61 @@ public String source() {
...
+                }
+            });
+    }
+
+    private <T> void safelyClusterStatePublished(ClusterStateTaskExecutor<T> executor, ClusterState newState) {
+        safely(
+            () -> executor.clusterStatePublished(newState),
+            e -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("exception thrown by executor while notifying of cluster state published from [{}]", e, newState.prettyPrint());
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: Are these debug level events? They all seem like bugs.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: ++ on error

**Comment by jasontedor** [comment]
Closing in favor of #15777.

**Pull request closed without merging**